chapter/4/dist-langs.md

```diff
@@ -4,7 +4,515 @@ title:  "Distributed Programming Languages"
+Objects that do not have a process are known as *passive* and more closely resemble traditional Java objects; their code is executed when called by processes belonging to other objects.
+Second, processes may not touch internal state (members) of other objects.
+Unlike Java, all internal state of Emerald objects must be accessed through method calls.
+Third, objects in Emerald may contain a special *monitor* section which can contain methods and variables that are accessed atomically.
+If multiple processes make simultaneous calls to a "monitored" method, the calls are effectively serialized.
+
+Emerald also takes an OOP approach to system upgrades.
+With a large system, it may not be desirable to disable the system, recompile, and re-launch.
+Emerald uses abstract types to define sets of interfaces.
+Objects that implement such interfaces can be "plugged in" where needed.
+Therefore, code may be dynamically upgraded, and different implementations may be provided for semantically similar operations.
+
+#### Argus (1988)
+
+Argus is a distributed programming language and system.
+It uses a special kind of object, called a *guardian*, to create units of distribution and group highly coupled data and logic.
+Argus procedures are encapsulated in atomic transactions, or *actions*, which allow operations that encompass multiple guardians to exhibit serializability.
+The presence of guardians and actions was motivated by Argus' use as a platform for building distributed, consistent applications.
+
+An object in Argus is known as a guardian.
+Like traditional objects, guardians contain internal data members for maintaining state.
+Unique to Argus is the distinction between volatile and stable data members.
+To cope with crashes, data members that are stable are periodically serialized to disk.
+When a guardian crashes and is restored, this serialized state is used to reconstruct guardians.
+Like in Emerald, internal data members may not be accessed directly, but rather through handlers.
+Guardians are interacted with through methods known as *handlers*.
+When a handler is called, a new process is created to handle the operation.
+Additionally, guardians may contain a background process for performing continual work.
+
+Argus encapsulates handler calls in what it calls *actions*.
+Actions are designed to solve the problems of consistency, synchronization, and fault tolerance.
+To accomplish this, actions are serializable as well as total.
+Being serializable means that no actions interfere with one another.
+For example, a read operation that spans multiple guardians either "sees" the complete effect of a simultaneous write operation, or it sees nothing.
+Being total means that write operations that span multiple guardians either fully complete or fully fail.
+This is accomplished by a two-phase commit protocol that serializes the state of *all* guardians involved in an action, or discards partial state changes.
```

Comment:
It's not only the commit protocol, it also relies on a complicated MVCC mechanism that operates cross node and has a garbage collection mechanism for removing stale versions.  This seems to be elided in this section and doesn't really expand on how Argus actually works.